llvm/lib/IR/DebugInfo.cpp

```diff
@@ -44,29 +44,12 @@ using namespace llvm;
 using namespace llvm::at;
 using namespace llvm::dwarf;
 
-void llvm::findDbgDeclares(SmallVectorImpl<DbgDeclareInst *> &DbgUsers,
-                           Value *V) {
-  // This function is hot. Check whether the value has any metadata to avoid a
-  // DenseMap lookup.
-  if (!V->isUsedByMetadata())
-    return;
-  auto *L = LocalAsMetadata::getIfExists(V);
-  if (!L)
-    return;
-  auto *MDV = MetadataAsValue::getIfExists(V->getContext(), L);
-  if (!MDV)
-    return;
+template <typename IntrinsicT, bool AnyType,
+          DPValue::LocationType Type = DPValue::LocationType(-1)>
```

Comment:
https://eel.is/c++draft/expr.static.cast para 10 indicates that this may be undefined behaviour. IMO we should bite the bullet and add an "End" type to the enumeration, followed by an "Any" type that's explicitly to be used in APIs like this. Slightly less than ideal, but at least well formed.
I'd also suggest "AnyKind" rather than "AnyType", to avoid mixing the ideas of C++ types and DPValue stuff.

Reply:
SGTM. Given it's already qualified by an enum class name `LocationType` I think `Any` should suffice? I'm not sure we need the `End` marker, but I've put it in there anyway. 

We could use bit flags (e.g. Any = 1 | 2)- I've no strong opinion either way and have not done so in the latest revsiion.

wdyt?